File: src/Practice/Spoj/Kquery.java
package Spoj;

/*
 * Created by bk on 21-12-2017 13:22
 *
5
5 1 2 3 4
3
2 4 1
4 4 4
1 5 2
 */

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

import static java.lang.Math.floorDiv;

class FastIO {
    private BufferedReader br;
    private StringTokenizer st;

    FastIO() {
        br = new BufferedReader(new
                InputStreamReader(System.in));
    }

    public String next() {
        while (st == null || !st.hasMoreElements()) {
            try {
                st = new StringTokenizer(br.readLine());
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return st.nextToken();
    }

    public int nextInt() {
        return Integer.parseInt(next());
    }

    public long nextLong() {
        return Long.parseLong(next());
    }

    public double nextDouble() {
        return Double.parseDouble(next());
    }

    public String nextLine() {
        String str = "";
        try {
            str = br.readLine();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return str;
    }
}

class Node {
    int min = 0, max = 0;

    @Override
    public String toString() {
        return "(" + this.min + ", " + this.max + ")";
    }
}

public class Kquery {
    static int[] arr;
    static Node[] segTree;

    public static void main(String[] __) {
        FastIO sc = new FastIO();
        int n = sc.nextInt();
        arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }
        segTree = new Node[4 * n];
        for (int i = 0; i < segTree.length; i++) segTree[i] = new Node();
        build(0, n - 1, 0);
        int q = sc.nextInt();
        for (int i = 0; i < q; i++) {
            int l = sc.nextInt() - 1;
            int r = sc.nextInt() - 1;
            int k = sc.nextInt() - 1;
            System.out.println(query(0, n - 1, l, r, 0, k));
        }
    }

    private static void build(int start, int end, int node) {
        if (start == end) {
            segTree[node].max = arr[start];
            segTree[node].min = arr[start];
        } else {
            int mid = floorDiv(start + end, 2);
            build(start, mid, 2 * node + 1);
            build(mid + 1, end, 2 * node + 2);
            segTree[node].min = Math.min(segTree[2 * node + 1].min, segTree[2 * node + 2].min);
            segTree[node].max = Math.max(segTree[2 * node + 1].max, segTree[2 * node + 2].max);
        }
    }

    // start l end r ; l start r end ; l start end r ; start l r end

    private static int query(int start, int end, int l, int r, int node, int k) {
        if ((l == start && r == end) || (l <= start && r <= end)) {
            if (segTree[node].min > k) return end - start + 1;
            else if (segTree[node].max < k) return 0;
            else {
                int mid = floorDiv(start + end, 2);
                return query(start, mid, l, r, 2 * node + 1, k) + query(mid + 1, end, l, r, 2 * node + 2, k);
            }
        } else if ((start <= l && end <= r) || (l <= start && r <= end) || (start <= l && r <= end)) {
            int mid = floorDiv(start + end, 2);
            return query(start, mid, l, r, 2 * node + 1, k) + query(mid + 1, end, l, r, 2 * node + 2, k);
        } else return 0;
    }
}

File: src/Practice/Spoj/kquery.py
"""16 8 7 4 3 2 1"""
from bisect import bisect_left as bs


class Node:
    lis = []

    def __init__(self, lis):
        self.lis = lis

    def __repr__(self):
        return str(self.lis)


def merge(l1, r1, l2, r2):
    i, j, k = l1, l2, l1
    temp = []
    while i <= r1 and j <= r2:
        if arr[i] < arr[j]:
            temp.append(arr[i])
            i += 1
        else:
            temp.append(arr[j])
            j += 1
    while i <= r1:
        temp.append(arr[i])
        i += 1
    while j <= r2:
        temp.append(arr[j])
        j += 1
    for i in range(len(temp)):
        arr[k] = temp[i]
        k += 1
    return temp


def mergeSort(arr, l, r, node):
    if l != r:
        mid = (l + r) // 2
        mergeSort(arr, l, mid, 2 * node + 1)
        mergeSort(arr, mid + 1, r, 2 * node + 2)
        t = merge(l, mid, mid + 1, r)
        segTree[node] = Node(t)
    else:
        segTree[node] = Node([arr[l]])


def query(l, r, start, end, k, node):
    cur = segTree[node].lis
    if l == start and r == end:
        return len(cur) - bs(cur, k + 0.1)
    elif l <= start <= end <= r:
        return len(cur) - bs(cur, k + 0.1)
    elif l <= start <= r <= end or start <= l <= end <= r or start <= l <= r <= end:
        mid = (start + end) // 2
        return query(l, r, start, mid, k, 2 * node + 1) + query(l, r, mid + 1, end, k, 2 * node + 1)
    else:
        return 0


if __name__ == '__main__':
    n = int(input().strip())
    arr = [int(i) for i in input().strip().split()]
    segTree = [Node([]) for i in range(4 * len(arr))]
    mergeSort(arr, 0, len(arr) - 1, 0)
    for i in range(int(input().strip())):
        l, r, k = [int(i) for i in input().strip().split()]
        l, r = l - 1, r - 1
        print(query(l, r, 0, n - 1, k, 0))
